src/System.IO.Pipelines/src/System/IO/Pipelines/PipeOperationState.cs

```diff
@@ -0,0 +1,76 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.IO.Pipelines
+{
+    [DebuggerDisplay("State: {_state}")]
+    internal struct PipeOperationState
+    {
+        private State _state;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void BeginRead()
+        {
+            if ((_state & State.BeginRead) == State.BeginRead)
+            {
+                ThrowHelper.ThrowInvalidOperationException_AlreadyReading();
+            }
+
+            _state &= ~State.EndRead;
+            _state |= State.BeginRead;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void BeginReadTentative()
+        {
+            if ((_state & State.BeginRead) == State.BeginRead)
+            {
+                ThrowHelper.ThrowInvalidOperationException_AlreadyReading();
+            }
+
+            _state &= ~State.EndRead;
+            _state |= State.BeginReadTenative;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void EndRead()
+        {
+            if ((_state & State.EndRead) == State.EndRead)
+            {
+                ThrowHelper.ThrowInvalidOperationException_NoReadToComplete();
+            }
+
+            _state &= ~(State.BeginRead | State.BeginReadTenative);
+            _state |= State.EndRead;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void BeginWrite()
+        {
+            _state |= State.WritingActive;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void EndWrite()
+        {
+            _state &= ~State.WritingActive;
+        }
+
+        public bool IsWritingActive => (_state & State.WritingActive) == State.WritingActive;
+
+        public bool IsReadingActive => (_state & State.BeginRead) == State.BeginRead;
+
+        [Flags]
+        internal enum State
+        {
+            BeginRead = 1,
+            BeginReadTenative = 2,
+            EndRead = 4,
```

Comment:
Why do you need `EndRead` state? The only check it is used for can be done via checking for `BeginRead`

Reply:
We have extremely weird behavior here.

- If you call TryRead and that returns false, then you call AdvanceTo it's supposed to work. In this case, nothing is set. This might be the one case that isn't possible to represent without a state. Just checking for the absence of BeginRead isn't enough.

Reply:
I just looked at this again, here's the state machine:

```
BeginRead()
EndRead()

OK

BeginReadTentative()
EndRead()

OK

EndRead() 

OK

BeginRead()
BeginRead() // Throws already reading

BeginRead()
BeginReadTentative() // Throws already reading

EndRead()
EndRead() // Throws no reading operation
```

Reply:
Why is a single `EndRead()` without beginread ok?